src/expectation/expectation.jl

```diff
@@ -0,0 +1,212 @@
+abstract type AbstractExpectationADAlgorithm  end
+struct NonfusedAD <: AbstractExpectationADAlgorithm end
+struct PrefusedAD <: AbstractExpectationADAlgorithm
+    norm_partials::Bool
+end
+PrefusedAD() = PrefusedAD(true)
+struct PostfusedAD <: AbstractExpectationADAlgorithm 
+    norm_partials::Bool
+end
+PostfusedAD() = PostfusedAD(true)
+
+abstract type AbstractExpectationAlgorithm <: DiffEqBase.DEAlgorithm end
+struct Koopman{TS} <:AbstractExpectationAlgorithm where TS<:AbstractExpectationADAlgorithm
+    sensealg::TS
+end
+Koopman() = Koopman(NonfusedAD())
+struct MonteCarlo <: AbstractExpectationAlgorithm 
+    trajectories::Int
+end
+
+# Builds integrand for arbitrary functions
+# TODO return in and out of place functions as tuple?
+function build_integrand(prob::ExpectationProblem, ::Koopman, ::Val{false})
+    @unpack g, d = prob
+    function(x,p)
+        g(x,p)*pdf(d,x)
+    end
+end
+
+# Builds integrand for DEProblems
+function build_integrand(prob::ExpectationProblem{F}, ::Koopman, ::Val{false}) where F<:SystemMap
+    @unpack S, g, h, d = prob
+    function(x,p)
+        ū, p̄ = h(x, p.x[1], p.x[2])
+        g(S(ū,p̄), p̄)*pdf(d,x)   
+    end
+end
+
+function _make_view(x::Union{Vector{T}, Adjoint{T, Vector{T}}}, i) where T
+    @view x[i]
+end
+
+function _make_view(x, i)
+    @view x[:,i]
+end
+
+function build_integrand(prob::ExpectationProblem{F}, ::Koopman, ::Val{true}) where F<:SystemMap
+    @unpack S, g, h, d = prob
+
+    if prob.nout == 1 # TODO fix upstream in quadrature, expected sizes depend on quadrature method is requires different copying based on nout > 1
+        set_result! = @inline function(dx, sol)
+            dx[:] .= sol[:]
```

Comment:
I'm not sure about the semantics of `sol[:]`, but it looks like there are unnecessary allocations from the `[:]` here.
Since they are marked `@inline`, I guess performance matters here?

Reply:
`set_result!` ends up formating the solution of an `EnsembleSolve` to the format required by Quadrature.jl. This is getting called once per batch integrand evaluation. I'm not sure how to clean up the allocation.

Reply:
This also highlights an interface issue with Quadrature (which I am probably guilty of creating), i.e. we shouldn't really need separate  `set_result!` for scalar and vector valued integrands.